### PR TITLE
BL-208 [goext] BeginTx() does not time out gracefully 

### PR DESCRIPTION
**Pull request by BartoszKakol**
Work in progress. Tests will be soon.

### CODE REVIEW

File: extension/goplugin/database.go [hunk 1]
```diff
@@ -64,7 +64,11 @@ func (db *Database) Begin() (goext.ITransaction, error) {
 // BeginTx starts a new transaction with options
 func (db *Database) BeginTx(ctx goext.Context, options *goext.TxOptions) (goext.ITransaction, error) {
 	opts := transaction.TxOptions{IsolationLevel: transaction.Type(options.IsolationLevel)}
-	t, err := db.raw.BeginTx(context.Background(), &opts)
+	newTimedContext := ctx.GetContext()
+	if _, ok := newTimedContext.Deadline();!ok{
```
Comment (zimnx) [inline extension/goplugin/database.go]: s/ok/hasDeadline

File: extension/goplugin/database.go [hunk 2]
```diff
@@ -64,7 +64,11 @@ func (db *Database) Begin() (goext.ITransaction, error) {
...
+		newTimedContext = context.WithTimeout(newTimedContext,5* time.Second)
```
Comment (zimnx) [inline extension/goplugin/database.go]: timeout should be configurable

Comment (p-kozlowski) [inline extension/goplugin/database.go]: please run goimports to format the code before pushing it. in GoLand you could add it in File->Tools->File Watchers - it will run whenever you save a file

Comment (BartoszKakol) [reply extension/goplugin/database.go]: It is configurable if you specify deadline in context. 5 seconds is only used if someone used function without specifying deadline.

Comment (zimnx) [reply extension/goplugin/database.go]: I meant that this 5 seconds value should be configurable. Although I don't think that's the right place to set this deadline. Maybe we should do at the beginning of request in api.go to make sure context is always there. What do you think?

Comment (BartoszKakol) [reply extension/goplugin/database.go]: We can return error from BeginTx if Timeout is not set, to force person using the function to add Timeout to Context. We need some sort of protection against Context without timeout otherwise we risk ending up with the same errors.

Comment (p-kozlowski) [reply extension/goplugin/database.go]: @BartoszKakol I think what @zimnx meant was that the timeout should be configurable using the config file, not by harcoding 5sec (and he's right about it). In my opinion that would be unnecessary - API requests always provide a context with timeout:
https://github.com/cloudwan/gohan/blob/master/extension/goplugin/environment.go#L438 (though again I agree it should be a part of Gohan core, not goplugin). timeout won't be here iff a user hand-mades a context, but that's already shooting oneself in their foot, so I wouldn't care that much for it.

in other words: just pass `ctx.GetContext()` to `BeginTx` and we'll be fine.

side note: returning an error when timeout is not set is not the best solution available. it's an API misuse and we should preferably make it a compile error, or, if that's not possible, `panic()` (aka. `assert` in other languages) to make it very clear that something is wrong.

File: extension/goplugin/database_test.go [hunk 3]
```diff
@@ -83,6 +85,20 @@ var _ = Describe("Database", func() {
 				Expect(err).To(HaveOccurred())
 				Expect(tx).To(BeNil())
 			})
+
+			It("should call inner function using Context with Timeout", func() {
+				tempCtxWithTimeout, _ := context.WithTimeout(context.Background(), 50*time.Millisecond)
+				ctx:= goext.Context{
```
Comment (p-kozlowski) [inline extension/goplugin/database_test.go]: `goimports` complain, but otherwise looks good

**Comment by BartoszKakol** [comment]
@morrisson Hi, this PR is approved, can you merge it, so we can close the ticket?

**Comment by p-kozlowski** [comment]
@BartoszKakol please squash the commits
